### PR TITLE
Improvements to Uniqueness Rule handling, and ACID improvements to agent merging

### PR DESCRIPTION
**Pull request by realVinayak**
1. Fixes https://github.com/specify/specify7/issues/3801. Checkout @melton-jason's comment (https://github.com/specify/specify7/issues/3801#issuecomment-1640975722) to do an initial test. 
2. Makes ACID improvements to merging. Left comments explaining those decisions.

Priorities: 

- [x] Instead of returning `http.HttpResponseNotFound` and `http.HttpResponseNotAllowed`, raise an exception, and then catch it from outside recursive merge. Otherwise it won't rollback at all. Currently reviewing changes to celery worker's part for this.
- [x] Take least ordering as new record in dedup. Added test for this also. Passes!
- [x] Add more comprehensive unit tests for some of the changes, and rollbacks.
- [ ] Fix this `# BUG: timestampmodified could be null for one record, and not the other`
- [ ] Fix uniqueness rule mess. Implementing solution in my comment there https://github.com/specify/specify7/issues/3841 
- [ ] Do code cleanups post merge with celery worker.

### CODE REVIEW

**Comment by grantfitzsimmons** [comment]
@realVinayak Can you provide some testing instructions so that @specify/ux-testing and I can test this?

File: specifyweb/businessrules/uniqueness_rules.py [hunk 1]
```diff
@@ -1,81 +1,180 @@
...
+       ]
+    }
+can simply be 
+    "AccessionAgent": [
+        (("role", "agent"), ("accession")),
+        (("role", "agent"), ("repositoryagreement"))
+        ]
+The second format also makes it much easier to construct django queries.
+So, parse_uniqueness_rules() automatically converts the current representation
+TODO: Refactor front-end to use this instead
```
Comment (maxxxxxdlp) [inline specifyweb/businessrules/uniqueness_rules.py:133]: @melton-jason just in case you didn't see this

File: specifyweb/specify/views.py [hunk 2]
```diff
@@ -105,19 +124,23 @@ def delete_blockers(request, model, id):
...
 @login_maybe_required
 @require_http_methods(['GET', 'HEAD'])
 def rows(request, model):
     "Returns tuples from the table for <model>."
     return api.rows(request, model)
 
+
 @require_http_methods(['GET', 'HEAD'])
-@cache_control(max_age=365*24*60*60, public=True)
+@cache_control(max_age=365 * 24 * 60 * 60, public=True)
```
Comment (maxxxxxdlp) [inline specifyweb/specify/views.py:143]: (optional) define global constants like `DAY`, `YEAR` ... and use them all around the back-end?

File: specifyweb/specify/views.py [hunk 3]
```diff
@@ -404,38 +449,75 @@ def set_admin_status(request, userid):
...
     """Replaces all the foreign keys referencing the old record ID
     with the new record ID, and deletes the old record.
     """
     # Confirm the target model table exists
     model_name = model_name.lower().title()
     target_model = getattr(spmodels, model_name)
     if target_model is None:
-        return http.HttpResponseNotFound("model_name: " + model_name + "does not exist.")
+        return http.HttpResponseNotFound(
+            "model_name: " + model_name + "does not exist.")
```
Comment (realVinayak) [inline specifyweb/specify/views.py:502]: BUG: Raise FailedMergingException here

**Comment by acwhite211** [comment]
I noticed a lot of new lines are formatting was done. Did you run a linter on the python code?  Maybe we should revert that and make new linting overall of files as a separate PR.

**Comment by acwhite211** [comment]
Everything else is looking good

**Comment by grantfitzsimmons** [comment]
@realVinayak Am I right to have closed this since #3846 has been merged?

**Comment by realVinayak** [comment]
@grantfitzsimmons yes

**Pull request closed without merging**